### ISSUE TITLE
`<complex>`: Implementation divergence for division by zero

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#2728]
Test case: https://godbolt.org/z/WqzW5b99E

```cpp
#include <complex>
#include <cstdio>
using namespace std;

int main() {
    using C = complex<double>;
    const C t1 = C{1.0} / 0.0;
    const C t2 = C{1.0} / C{0.0};
    printf("    t1: (%g, %g)\n", t1.real(), t1.imag());
    printf("t1 alt: (%g, -%g)\n", t1.real(), -t1.imag());
    printf("    t2: (%g, %g)\n", t2.real(), t2.imag());
}
```

MSVC has always printed (this is not a regression):
```
    t1: (inf, -nan(ind))
t1 alt: (inf, -nan)
    t2: (nan, nan)
```

GCC/libstdc++ and Clang/libc++ print:

```
    t1: (inf, -nan)
t1 alt: (inf, -nan)
    t2: (inf, -nan)
```

This was originally reported as DevCom-10040631 and internal VSO-1539658 / AB#1539658 . The user is wondering whether this is a bug.

I've added "t1 alt" to show that "-nan(ind)" is just how MSVC likes to print negative NaN. Otherwise, the behavior for `C{1.0} / 0.0` is consistent across implementations, and our implementation is just component-wise, so this is unsurprising:

https://github.com/microsoft/STL/blob/1a20fe1133d711a647bbb135d98743f91b7be323/stl/inc/complex#L1465-L1471

For `C{1.0} / C{0.0}`, we go through:

https://github.com/microsoft/STL/blob/1a20fe1133d711a647bbb135d98743f91b7be323/stl/inc/complex#L993-L995

within the `_Div()` function:

https://github.com/microsoft/STL/blob/1a20fe1133d711a647bbb135d98743f91b7be323/stl/inc/complex#L970-L1009

Therefore we're deliberately returning `(nan, nan)` here.

**Comment by StephanTLavavej** [comment]
I'm not sure what the Standard says - is this covered by WG21-N4910 [[complex.numbers.general]/3](https://eel.is/c++draft/complex.numbers.general#3) "If the result of a function is not mathematically defined or not in the range of representable values for its type, the behavior is undefined."?

**Comment by fsb4000** [comment]
I found that clang changed their behaviour in clang 3.6,
gcc always has current behaviour:
https://godbolt.org/z/E6zo5T394

clang 3.5.2:
```
    t1: (-nan, -nan)
t1 alt: (-nan, -nan)
    t2: (-nan, -nan)
```
clang 3.6:
```
    t1: (inf, -nan)
t1 alt: (inf, -nan)
    t2: (inf, -nan)
```

**Comment by statementreply** [comment]
**TL;DR:** `(NaN, NaN)` is probably wrong.

While it is not explicitly stated for `operator/`, I think that the intention is to what C does for complex numbers if applicable. This can be seen from "_Remarks_: Behaves the same as the C function `cmeow`. See also: ISO C x.x.x.x" in [[complex.transcendentals]](https://eel.is/c++draft/complex.transcendentals) and the discussion of LWG-2597.

Annex G (IEC 60559-compatible complex arithmetic) of the C standard specifies the behavior regarding to IEEE-754 special values and exceptions.

[G.3]/1
> A complex or imaginary value with at least one infinite part is regarded as an _infinity_ (even if its
other part is a quiet NaN). A complex or imaginary value is a _finite number_ if each of its parts is a
finite number (neither infinite nor NaN). A complex or imaginary value is a _zero_ if each of its parts is
a zero.

[G.5.1]/4:4
> if the first operand is a nonzero finite number or an infinity and the second operand is a zero,
then the result of the `/` operator is an infinity.

If we follow C's behavior of the `/` operator, the result of `(1.0, 0.0) / (0.0, 0.0)` should be a complex infinity (at least one of `real()` and `imag()` is infinite), not `(NaN, NaN)`.

The C standard doesn't say which infinity to return. The basic idea behind the design is that, even for zeroes and infinities, `abs(z / w) = abs(z) / abs(w)` and `arg(z / w) = arg(z) - arg(w)` still approximately holds to some very limited precision. There are controversies and implementation divergence on this (e.g. glibm returns `(Inf, NaN)`, Matlab returns `(Inf, 0)`).

See also #887.
